org/mozilla/jss/ssl/SSLCipher.c

```diff
@@ -6,26 +6,112 @@
 
 #include "_jni/org_mozilla_jss_ssl_SSLCipher.h"
 
+/* Copied from NSS's ssl3con.c. */
+static const CK_MECHANISM_TYPE auth_alg_defs[] = {
+    CKM_INVALID_MECHANISM, /* ssl_auth_null */
+    CKM_RSA_PKCS,          /* ssl_auth_rsa_decrypt */
+    CKM_DSA, /* ? _SHA1 */ /* ssl_auth_dsa */
+    CKM_INVALID_MECHANISM, /* ssl_auth_kea (unused) */
+    CKM_ECDSA,             /* ssl_auth_ecdsa */
+    CKM_ECDH1_DERIVE,      /* ssl_auth_ecdh_rsa */
+    CKM_ECDH1_DERIVE,      /* ssl_auth_ecdh_ecdsa */
+    CKM_RSA_PKCS,          /* ssl_auth_rsa_sign */
+    CKM_RSA_PKCS_PSS,      /* ssl_auth_rsa_pss */
+    CKM_NSS_HKDF_SHA256,   /* ssl_auth_psk (just check for HKDF) */
+    CKM_INVALID_MECHANISM  /* ssl_auth_tls13_any */
+};
+PR_STATIC_ASSERT(PR_ARRAY_SIZE(auth_alg_defs) == ssl_auth_size);
+
+/* Copied from NSS's ssl3con.c. */
+static const CK_MECHANISM_TYPE kea_alg_defs[] = {
+    CKM_INVALID_MECHANISM, /* ssl_kea_null */
+    CKM_RSA_PKCS,          /* ssl_kea_rsa */
+    CKM_DH_PKCS_DERIVE,    /* ssl_kea_dh */
+    CKM_INVALID_MECHANISM, /* ssl_kea_fortezza (unused) */
+    CKM_ECDH1_DERIVE,      /* ssl_kea_ecdh */
+    CKM_ECDH1_DERIVE,      /* ssl_kea_ecdh_psk */
+    CKM_DH_PKCS_DERIVE,    /* ssl_kea_dh_psk */
+    CKM_INVALID_MECHANISM, /* ssl_kea_tls13_any */
+};
+PR_STATIC_ASSERT(PR_ARRAY_SIZE(kea_alg_defs) == ssl_kea_size);
+
+#if (NSS_VMAJOR >= 3) && (NSS_VMINOR >= 43)
+/* Not present in ssl3con.c. */
+static const CK_MECHANISM_TYPE hash_alg_defs[] = {
+    CKM_INVALID_MECHANISM, /* ssl_hash_none */
+    CKM_MD5,               /* ssl_hash_md5 */
+    CKM_SHA_1,             /* ssl_hash_sha1 */
+    CKM_SHA224,            /* ssl_hash_sha224 */
+    CKM_SHA256,            /* ssl_hash_sha256 */
+    CKM_SHA384,            /* ssl_hash_sha384 */
+    CKM_SHA512,            /* ssl_hash_sha512 */
+};
+#endif
+
 JNIEXPORT jboolean JNICALL
 Java_org_mozilla_jss_ssl_SSLCipher_checkSupportedStatus(JNIEnv *env, jclass clazz, jint cipher_suite)
 {
     PRInt32 allowed;
+    SSLCipherSuiteInfo info = { 0 };
+    int have_auth_type;
+    int have_kea_type;
 
+    /* Fetch information about whether or not this cipher is allowed by local
+     * policy. */
     if (SSL_CipherPolicyGet(cipher_suite, &allowed) != SECSuccess) {
         return JNI_FALSE;
     }
 
-    if (!PK11_IsFIPS() || allowed != SSL_ALLOWED) {
-        return (allowed == SSL_ALLOWED) ? JNI_TRUE : JNI_FALSE;
+    /* Fetch extended information about this particular cipher suite. */
+    if (SSL_GetCipherSuiteInfo(cipher_suite, &info, sizeof(info)) != SECSuccess) {
+        return JNI_FALSE;
+    }
+
+    have_auth_type = info.length >= (((void *)&info.authType - (void *)&info) + sizeof(info.authType));
+    have_kea_type = info.length >= (((void *)&info.keaType - (void *)&info) + sizeof(info.keaType));
+
+    if (allowed != SSL_ALLOWED) {
+        /* If the cipher suite isn't allowed by policy, reject it early. */
+        return JNI_FALSE;
     }
```

Comment:
This behavior is intentional.

If the policy doesn't allow a cipher suite (and in the unlikely case where the NSS DB is in FIPS mode but the system isn't), we shouldn't negotiate that cipher suite. There's a few overlapping items here that make this more comprehensive of a "supported" check:

 - Does local policy allow us to use this cipher?
 - Can we use this cipher in FIPS mode (if the system or NSS DB is in it)?
 - Do we have all of the mechanisms necessary to support this cipher?

Reply:
OK.

Reply:
I agree... I think we could move this up though.

In particular, if `SSL_CipherPolicyGet` fails, it shows that we've passed an invalid cipher suite (which shouldn't happen in most cases, unless we add a new one to the `SSLCipher` enum and the underlying NSS doesn't support it). We can then check allowed status and return `JNI_FALSE` if it isn't allowed at all. Then when `SSL_GetCipherSuiteInfo` is called -- the only failure case remaining is when the `cipher_suite` struct we pass exceeds the size of what the system NSS library supports. We can return `JNI_TRUE` instead of `JNI_FALSE` -- because it indicates an internal error and we can't get more precise information.